### DISCUSSION TITLE
Compress the generated Visio files.

### DISCUSSION THREAD

**Discussion by vidhanbagdai** [clientIO/joint#1806] [Q&A]
As from the below document, the toVSDX function takes compression parameter to reduce the size of the file and as mentioned with more compression level the time is expected to increase, but there were no benchmarks mentioned
src: https://resources.jointjs.com/docs/rappid/v3.5/format.Visio.html#VisioArchive.prototype.toVSDX

We performed some simple runs across a JointJS graph with 100 nodes

Observations:
<html>
<body>

Compression level | Size | Avg Time (s)
-- | -- | --
No compression | 14 mb | 22.265
1 | 3.4mb | 21.299
3 | 3.1mb | 21.765
5 | 2.8mb | 21.121
7 | 2.6mb | 20.496
9 | 2.6mb | 20.369

</body>
</html>

min time: ~19s
max time: ~32s
Memory spike: 480MB

As expected, higher compression levels result in better file sizes. But the time taken for compression is not consistent.
Also the time varies each time the server is used, sometimes it takes less than 10s and sometimes more than 30s.

Queries:

1. Can any details be provided regarding the compression algorithm and how to get expected results?
2. Can the memory consumption be reduced?
3. If not, can the file be split, the parts compressed and then be assembled back as one?

**Comment by kumilingus** [comment]
The compression of the `vsdx` file is done through `jszip` library. We use the [generateAsync](https://stuk.github.io/jszip/documentation/api_jszip/generate_async.html) method specifically and pass the options to it unchanged. You should be able to find your answers there.

Using the [generateInternalStream](https://stuk.github.io/jszip/documentation/api_jszip/generate_internal_stream.html) might help.
```js
archive.zip.generateInternalStream(/* ... */);
```
